File: backend/src/services/GameService.ts
import { randomUUID } from "node:crypto";
import type {
  ActionRequest,
  GameEvent,
  GameStateDTO,
  Orientation,
  ShipKind,
} from "@shared/index.js";
import { isEndTurnRequest } from "@shared/index.js";
import { AIOpponent } from "../ai/AIOpponent.js";
import { ApiError } from "../api/errors.js";
import { Game } from "../core/Game.js";
import { Rng } from "../core/rng.js";
import { toGameStateDTO } from "../core/dto.js";
import { GameStore } from "./GameStore.js";

/**
 * Use-case layer. Controllers only call methods on this class.
 *
 * Turn model:
 *   - Human acts via POST /actions (one ship action OR endTurn).
 *   - On endTurn, we transition activePlayer to "ai" but DO NOT play the AI.
 *   - The client then polls POST /ai-step, which executes one AI action per
 *     call and returns its events. When the AI is out of actions, control
 *     transitions back to "human" automatically.
 *
 * Each AI ship has a per-game `AIOpponent` instance keyed by the same RNG
 * stored on the Game. Reusing a fresh AIOpponent each step is harmless
 * because all AI state is derived from the public board view.
 */
export class GameService {
  constructor(private store: GameStore = new GameStore()) {}

  // ----- Lifecycle -----------------------------------------------------

  createGame(seed?: number): GameStateDTO {
    const rng = seed != null ? new Rng(seed) : Rng.random();
    const game = new Game(randomUUID(), rng);
    const ai = new AIOpponent(rng);
    ai.placeFleet(game);
    this.store.create(game);
    return toGameStateDTO(game, "human");
  }

  getView(id: string): GameStateDTO {
    return toGameStateDTO(this.requireGame(id), "human");
  }

  deleteGame(id: string): void {
    this.store.delete(id);
  }

  // ----- Placement ----------------------------------------------------

  placeShip(
    id: string,
    kind: ShipKind,
    anchor: number,
    orientation: Orientation,
  ): GameStateDTO {
    const game = this.requireGame(id);
    game.placeHumanShip(kind, anchor, orientation);
    return toGameStateDTO(game, "human");
  }

  placeRandom(id: string): GameStateDTO {
    const game = this.requireGame(id);
    game.placeHumanRandom();
    return toGameStateDTO(game, "human");
  }

  resetPlacement(id: string): GameStateDTO {
    const game = this.requireGame(id);
    game.resetHumanPlacement();
    return toGameStateDTO(game, "human");
  }

  // ----- Play ---------------------------------------------------------

  applyAction(
    id: string,
    req: ActionRequest,
  ): { state: GameStateDTO; events: GameEvent[] } {
    const game = this.requireGame(id);
    if (game.isOver()) {
      throw new ApiError("GAME_OVER", "Game has ended.");
    }
    if (game.phase !== "playing") {
      throw new ApiError("WRONG_PHASE", "Actions only allowed during play.");
    }

    let events: GameEvent[] = [];
    if (isEndTurnRequest(req)) {
      events = game.endHumanTurnTransition();
    } else {
      events = game.applyHumanAction(req);
      // Auto-end the turn for the human when they have no actions left.
      if (
        !game.isOver() &&
        game.activePlayer === "human" &&
        !game.humanHasActionsLeft()
      ) {
        events = events.concat(game.endHumanTurnTransition());
      }
    }
    return { state: toGameStateDTO(game, "human"), events };
  }

  /**
   * Execute a single AI action. If the AI has nothing left, transitions back
   * to the human and returns the `turn_started` event for the human.
   *
   * Idempotent-ish: if called when it's NOT the AI's turn, returns the
   * current state with empty events.
   */
  aiStep(id: string): { state: GameStateDTO; events: GameEvent[]; aiDone: boolean } {
    const game = this.requireGame(id);
    if (game.isOver()) {
      return {
        state: toGameStateDTO(game, "human"),
        events: [],
        aiDone: true,
      };
    }
    if (game.activePlayer !== "ai") {
      return {
        state: toGameStateDTO(game, "human"),
        events: [],
        aiDone: true,
      };
    }
    const ai = new AIOpponent(game.rng);
    const stepEvents = ai.stepOnce(game);
    const finishEvents = game.finishAiTurnIfDone();
    const aiDone = game.activePlayer !== "ai" || game.isOver();
    return {
      state: toGameStateDTO(game, "human"),
      events: [...stepEvents, ...finishEvents],
      aiDone,
    };
  }

  // ----- helpers ------------------------------------------------------

  private requireGame(id: string): Game {
    const game = this.store.get(id);
    if (!game) throw new ApiError("NOT_FOUND", `No game with id ${id}`);
    return game;
  }
}

File: backend-rs/src/services/game_service.rs
use uuid::Uuid;

use crate::ai::AIOpponent;
use crate::api::errors::{ApiError, ApiErrorCode, ApiResult};
use crate::core::dto::to_game_state_dto;
use crate::core::game::Game;
use crate::core::rng::Rng;
use crate::shared::{
    ActionRequest, GameEvent, GameStateDTO, Orientation, PlayerSide, ShipKind,
};

use super::store::GameStore;

/// Use-case layer. Controllers only call methods on this class.
#[derive(Clone)]
pub struct GameService {
    pub store: GameStore,
}

impl GameService {
    pub fn new(store: GameStore) -> Self {
        Self { store }
    }

    // ----- Lifecycle ----------------------------------------------------

    pub fn create_game(&self, seed: Option<u32>) -> GameStateDTO {
        let mut rng = match seed {
            Some(s) => Rng::new(s),
            None => Rng::random(),
        };
        let id = Uuid::new_v4().to_string();
        let mut game = Game::new(id, rng.clone());
        // Place the AI fleet using a clone of the RNG; advance the game's
        // canonical RNG to match so subsequent random ops are deterministic
        // given a seed (mirrors the TS code, which shares one Rng instance).
        AIOpponent::place_fleet(&mut game, &mut rng);
        game.rng = rng;
        let state = to_game_state_dto(&game, PlayerSide::Human);
        self.store.create(game);
        state
    }

    pub fn get_view(&self, id: &str) -> ApiResult<GameStateDTO> {
        let g = self.require_game(id)?;
        let g = g.lock();
        Ok(to_game_state_dto(&g, PlayerSide::Human))
    }

    pub fn delete_game(&self, id: &str) {
        self.store.delete(id);
    }

    // ----- Placement ----------------------------------------------------

    pub fn place_ship(
        &self,
        id: &str,
        kind: ShipKind,
        anchor: u32,
        orientation: Orientation,
    ) -> ApiResult<GameStateDTO> {
        let g = self.require_game(id)?;
        let mut g = g.lock();
        g.place_human_ship(kind, anchor, orientation)?;
        Ok(to_game_state_dto(&g, PlayerSide::Human))
    }

    pub fn place_random(&self, id: &str) -> ApiResult<GameStateDTO> {
        let g = self.require_game(id)?;
        let mut g = g.lock();
        g.place_human_random()?;
        Ok(to_game_state_dto(&g, PlayerSide::Human))
    }

    pub fn reset_placement(&self, id: &str) -> ApiResult<GameStateDTO> {
        let g = self.require_game(id)?;
        let mut g = g.lock();
        g.reset_human_placement()?;
        Ok(to_game_state_dto(&g, PlayerSide::Human))
    }

    // ----- Play ---------------------------------------------------------

    pub fn apply_action(
        &self,
        id: &str,
        req: ActionRequest,
    ) -> ApiResult<(GameStateDTO, Vec<GameEvent>)> {
        let g = self.require_game(id)?;
        let mut g = g.lock();
        if g.is_over() {
            return Err(ApiError::new(ApiErrorCode::GameOver, "Game has ended."));
        }
        if g.phase != crate::shared::GamePhase::Playing {
            return Err(ApiError::new(
                ApiErrorCode::WrongPhase,
                "Actions only allowed during play.",
            ));
        }

        let events: Vec<GameEvent> = match req {
            ActionRequest::EndTurn => g.end_human_turn_transition()?,
            ActionRequest::Ship(ship_req) => {
                let mut ev = g.apply_human_action(&ship_req)?;
                if !g.is_over()
                    && g.active_player == PlayerSide::Human
                    && !g.human_has_actions_left()
                {
                    let mut more = g.end_human_turn_transition()?;
                    ev.append(&mut more);
                }
                ev
            }
        };
        let state = to_game_state_dto(&g, PlayerSide::Human);
        Ok((state, events))
    }

    /// Execute a single AI action.
    pub fn ai_step(&self, id: &str) -> ApiResult<(GameStateDTO, Vec<GameEvent>, bool)> {
        let g = self.require_game(id)?;
        let mut g = g.lock();
        if g.is_over() {
            return Ok((to_game_state_dto(&g, PlayerSide::Human), Vec::new(), true));
        }
        if g.active_player != PlayerSide::Ai {
            return Ok((to_game_state_dto(&g, PlayerSide::Human), Vec::new(), true));
        }
        let step_events = AIOpponent::step_once(&mut g);
        let finish_events = g.finish_ai_turn_if_done();
        let ai_done = g.active_player != PlayerSide::Ai || g.is_over();
        let mut events = step_events;
        events.extend(finish_events);
        Ok((to_game_state_dto(&g, PlayerSide::Human), events, ai_done))
    }

    // ----- helpers ------------------------------------------------------

    fn require_game(
        &self,
        id: &str,
    ) -> ApiResult<std::sync::Arc<parking_lot::Mutex<Game>>> {
        self.store.get(id).ok_or_else(|| {
            ApiError::new(ApiErrorCode::NotFound, format!("No game with id {}", id))
        })
    }
}
